jena-core/src/main/java/org/apache/jena/graph/GraphMemFactory.java

```diff
@@ -30,8 +30,17 @@
 /**
  * A factory class for creating memory Graphs.
  * <p>
- * Apache Jena is migrating to term semantics graph for consistency across all in-memory and persistent storage graphs
- *
+ * Apache Jena is migrating to term semantics graph for consistency across all in-memory and persistent storage graphs.
+ * <p>
+ * All the graphs that this factory creates are <strong>not thread-safe</strong>.
+ * Note that if the memory Graph is structurally modified at any time after
+ * the iterator has been created by any of the {@code find*} or {@code stream*} methods, the iterator may throw
+ * a {@link java.util.ConcurrentModificationException ConcurrentModificationException}
+ * if continued with it after this modification.
+ * This may happen even if the queried data does not relate directly to modified data
+ * (i.e. when triple search pattern does not match added or deleted triple).
+ * The good practice is to explicitly close iterators immediately after a read operation or
+ * use terminate operations such as {@link ExtendedIterator#toList()}.
```

Comment:
```suggestion
 * use materialising operations such as {@link ExtendedIterator#toList()}.
```

Reply:
In java-stream API such operations are called `terminal`, so it's typo. 

maybe `materializing (or terminal, in terms of Java Stream API)` ?

Reply:
"terminal" is slightly different - it does not necessary read the whole stream (e.g. `findFirst`) but there is an internal "steam close" call .  

`ExtendedIterator` does not this hidden close for partially consumed iterators. It's run-to-end or call `ExtendedIterator.close`. `materialising` is running to the end.

Reply:
When we call `Stream.findFirst`, we can't use this stream anymore, so for Graph-mem it seems safe even without explicit close. Although this is a good practice.

As far as I can see, there is still no `org.apache.java.util.iterator.ExtendedIterator.findFirst` method and other terminal methods. Such operations are available only in the form of utilities (`org.apache.java.util.iterator.Iter.findFirst`) and they do not call close. But, by the way, it makes them not so useful.

The existing "ExtendedIterator" materializing methods are also terminal. So I don't see any logical error here.

In addition, `Graph` now supports Java Streams, so we also have to say a few words about Streams.

I think some word could be added directly to `Graph` interface, and additional explanations could be added to `GraphMemFactory` as well, to make it less ambiguous.

By the way, why we use `org.apache.jena.atlas.lib.Closeable` instead of `java.lang.AutoCloseable`?
I think we can do something like this:
```java
interface org.apache.jena.atlas.lib.Closeable extends java.lang.AutoCloseable {

    @Override
    void close(); // no exception
}
```
\+ In ONTAPI we also have utils (com.github.sszuev.jena.ontapi.utils.Iterators)[https://github.com/sszuev/jena-owl2/blob/main/src/main/java/com/github/sszuev/jena/ontapi/utils/Iterators.java].
What do you think about adding these functionality to Jena's `Iter`?

Reply:
`AutoCloseable.close` is required. Jena allows exhausting the iterator.

Reply:
It looks like I still don't quite understand the functional difference between `ExtendedIterator` and `Stream`. 

Both closeable, both have terminal and intrermediate operations (e.g. intermediate: `ExtendedIterator.filterKeep`, `Stream.filter`), both can be exhausted or not), both lazy. 
What am I missing out on?

`AutoCloseable` is not required, it is also "good practice", but in Java we can use `try-with-resources` to make our code safer. 
Also, switching to `AutoCloseable` will not break user code.